junebug/workers.py

```diff
@@ -33,27 +44,76 @@ class MessageForwardingWorker(ApplicationWorker):
         self.redis = yield TxRedisManager.from_config(
             self.config['redis_manager'])
-        self.message_store = InboundMessageStore(
-            self.redis, self.config['ttl'])
+
+        self.inbounds = InboundMessageStore(
+            self.redis, self.config['inbound_ttl'])
+
+        self.outbounds = OutboundMessageStore(
+            self.redis, self.config['outbound_ttl'])
 
     @inlineCallbacks
     def teardown_application(self):
         yield self.redis.close_manager()
 
+    @property
+    def channel_id(self):
+        return self.config['transport_name']
+
     @inlineCallbacks
     def consume_user_message(self, message):
         '''Sends the vumi message as an HTTP request to the configured URL'''
-        config = yield self.get_config(message)
-        yield self.message_store.store_vumi_message(
-            config.transport_name, message)
-        url = config.mo_message_url.encode('utf-8')
-        headers = {
-            'Content-Type': 'application/json',
-        }
-        msg = json.dumps(
-            Channel.api_from_message(message), cls=JSONMessageEncoder)
-        resp = yield treq.post(url, data=msg, headers=headers)
-        if resp.code < 200 or resp.code >= 300:
+        yield self.inbounds.store_vumi_message(self.channel_id, message)
+
+        msg = api_from_message(message)
+        resp = yield post(self.config['mo_message_url'], msg)
+
+        if request_failed(resp):
             logging.exception(
                 'Error sending message, received HTTP code %r with body %r. '
                 'Message: %r' % (resp.code, (yield resp.content()), msg))
+
+    @inlineCallbacks
+    def forward_event(self, event):
+        url = yield self._get_event_url(event)
+
+        if url is None:
+            logging.exception(
+                "Outbound message not found for event %r" % (event,))
+            return
+
+        msg = api_from_event(self.channel_id, event)
+
+        if msg['event_type'] is None:
+            logging.exception("Discarding unrecognised event %r" % (event,))
+            return
+
+        resp = yield post(url, msg)
+
+        if request_failed(resp):
+            logging.exception(
+                'Error sending event, received HTTP code %r with body %r. '
+                'Event: %r' % (resp.code, (yield resp.content()), event))
+
+    def consume_ack(self, event):
+        return self.forward_event(event)
+
+    def consume_nack(self, event):
+        return self.forward_event(event)
+
+    def consume_delivery_report(self, event):
+        return self.forward_event(event)
+
+    def _get_event_url(self, event):
+        msg_id = event['user_message_id']
+        return self.outbounds.load_event_url(self.channel_id, msg_id)
+
+
+def request_failed(resp):
+    return resp.code < 200 or resp.code >= 300
+
+
+def post(url, data):
+    return treq.post(
+        url.encode('utf-8'),
+        data=json.dumps(data, cls=JSONMessageEncoder),
```

Comment:
Not for this PR, but I'd like to avoid using Vumi's `JSONMessageEncoder` because it felt like that was a mistake we made early on in Vumi that was difficult to undo later (i.e. running through everything magically handling datetimes).